### PR TITLE
Pad using weighted vertex normals

### LINKED ISSUE #76: Non-uniform padding of non-convex, long meshes
I noticed that when applying a padding factor to non-convex, long-in-one dimension meshes, such as the table pictured below, the padding applied to the vertices is not uniform, since it is applied along the vector joining the mesh center and the vertex being padded. Any advice on how to make the padding more accurate for this kind of meshes?

Test case example: I have a robot arm on top of the table depicted below and want to add padding of 10 cm to the table to avoid the robot hand to collide during pick and place, meaning that I'd expect to have the table's top, for instance, "grow up" 10 cm (besides "growing in the X and Y directions as well).

In reality, when I apply padding to the table, what I get is that the table top grows in X ~9cm, whereas in Y and Z it grows only ~3cm.

table

### PR DESCRIPTION
**Pull request by kbrameld**
Resolves: #76

As discussed in the linked issue, padding behavior for complex meshes are not good. These changes improve upon the padding functionality, and obtain a closer functionality to blender's fatten method. Two main changes:
* padding is performed in direction of vertex normal instead of direction from center of mesh
* calculates better weighted vertex normal by accounting for the angle of a face at a vertex.

A complex STL file (original) | Current padding method
:----:|:----:
Screenshot from 2024-04-02 15-01-38 |  Screenshot from 2024-04-02 14-01-02

Fatten method (in Blender) | New padding method
:----:|:----:
Screenshot from 2024-04-02 11-39-09 | Screenshot from 2024-04-02 14-27-05

Non-uniform padding x-direction | y-direction | z-direction
:----:|:----:|:----:
Screenshot from 2024-04-02 15-00-00 |  Screenshot from 2024-04-02 15-32-50 | Screenshot from 2024-04-02 15-32-27

As discussed in the linked issue, although this is a general improvement, this change may potentially affect behavior in downstream packages if they relied on the bad padding.

### CODE REVIEW

**Comment by peci1** [comment]
Wow, it's been a long time since I was interested in this issue.

In #127, I found:

> Another comment was that generally change of the current behavior is okay as long as we can provide some guarantees on the relationship between the old approach and the new one (i.e. all points contained by the old behavior would be contained by the new one, or so...).

Could you sketch up a "proof" of some relationship?

Or, alternatively, wouldn't it be easier to create a function with a new name (`fatten`?) instead of `scaleAndPad` to retain the original behavior for existing code?

**Comment by kbrameld** [comment]
> Another comment was that generally change of the current behavior is okay as long as we can provide some guarantees on the relationship between the old approach and the new one (i.e. all points contained by the old behavior would be contained by the new one, or so...).

Not sure if there is a nice relationship. The old padding method was closer to a scaling operation, and the resulting mesh often didn't include the original vertices. Here's an example with the blue being the padded shapes:

Old method | New Method
:----:|:----:
Screenshot from 2024-03-14 14-54-19 | Screenshot from 2024-03-14 14-46-05

The only meshes that will look similar between the old and new method will be meshes where the weighted vertex normal direction is close to the direction of the vertex from the center. (eg. cube, sphere meshes) For any other mesh, the padded shapes will look different.

> Or, alternatively, wouldn't it be easier to create a function with a new name (fatten?) instead of scaleAndPad to retain the original behavior for existing code?

Yes, this can be considered an alternative. If we go this approach, I advocate for deprecating the "pad" methods (in upcoming ros2 distros).

File: src/shapes.cpp
Comment (rhaschke) [inline src/shapes.cpp]: Consider using `Eigen::Map<Eigen::Vector3d>(vertices + 3*vi, 3)` to avoid copying the data.

File: src/shapes.cpp
Comment (rhaschke) [inline src/shapes.cpp]: Consider using an `Eigen::Map<Eigen::MatrixXd>` to `vertex_normals` (a vertex_count x 3 matrix) to avoid allocating an extra array (and final data copying).

File: src/shapes.cpp
Comment (rhaschke) [inline src/shapes.cpp]: As far as I see, this method replaces the old method `computeVertexNormals()`.

File: src/shapes.cpp
Comment (rhaschke) [inline src/shapes.cpp]: Don't you need to handle the special case |v1| = 0 or |v2| = 0, i.e. dividing zero by zero?

**Review by rhaschke — COMMENTED** [review]
Thanks for this contribution.
Given that the old mesh padding was not working as expected, I don't see the need to introduce the new behavior with a new name. I even suggest replacing the old method `computeVertexNormals()` with the new one.

**Comment by peci1** [comment]
> Given that the old mesh padding was not working as expected, I don't see the need to introduce the new behavior with a new name.

At least robot_body_filter users actively use padding with its current behavior. Any big changes could break their workflows. But as long as the result is similar for the commonly found meshes (e.g. UR-5), it should be okay. I think the example mesh used in this PR is kind of extreme and very far from what people normally use, so the big difference in the example is okay.

**Comment by peci1** [comment]
@kbrameld Do I get it correctly that the unexpected case of enlarging a square would still be present with this PR? https://github.com/ros-planning/geometric_shapes/issues/127#issue-587387573 .

File: src/shapes.cpp
Comment (kbrameld) [reply src/shapes.cpp]: I like this suggestion. To match data storage order, either we need a 3xN or Nx3 (with Row major). Not sure if there is much difference to the two, so I'm going to go with ``Eigen::Map<Eigen::Matrix<double, 3, Eigen::Dynamic>>``.

File: src/shapes.cpp
Comment (kbrameld) [reply src/shapes.cpp]: Changed to replace current method rather than adding a new method

File: src/shapes.cpp
Comment (kbrameld) [reply src/shapes.cpp]: Good catch, updated

**Comment by kbrameld** [comment]
> @kbrameld Do I get it correctly that the unexpected case of enlarging a square would still be present with this PR? [#127 (comment)](https://github.com/ros-planning/geometric_shapes/issues/127#issue-587387573) .

You're correct, the changes in this PR won't change the behavior of enlarging the square in that issue.

File: src/shapes.cpp
Comment (rhaschke) [inline src/shapes.cpp]: If you Eigen::Map `triangle_normals` as well, this code further simplifies to vector addition.

Comment (kbrameld) [reply src/shapes.cpp]: Sounds good, addressed

**Review by rhaschke — APPROVED** [review]

**Comment by ijnek** [comment]
Once this is merged, I will create a PR against the ``ros2`` branch too. What's required to move forward with this PR?

**Comment by ijnek** [comment]
Pinging this thread for attention to be reviewed

**Comment by rhaschke** [comment]
Merging this, w/o a second review as nobody seems to do it...

**Comment by kbrameld** [comment]
Thanks for reviewing @rhaschke!

**Comment by peci1** [comment]
I've tested robot_body_filter (which uses geometric_shapes) with a few models from https://github.com/ctu-vras/rosdevday_cloud_filtering and the models seem to be affected only a little by this PR. So it seems this change should be safe for most users! Thanks.

**Comment by timonegk** [comment]
> Once this is merged, I will create a PR against the `ros2` branch too.

Now that this is merged, could you create a PR for ROS 2?

**Pull request merged**
